osu.Game/Rulesets/Mods/IApplicableRestartOnFail.cs

```diff
@@ -0,0 +1,16 @@
+// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+namespace osu.Game.Rulesets.Mods
+{
+    /// <summary>
+    /// Represents a mod which can request to restart on fail.
+    /// </summary>
+    public interface IApplicableRestartOnFail : IApplicableMod
```

Comment:
How about combining this into `IApplicableFailOverride`? Make a property like `RestartOnFail` in it.